HW5.py

```diff
@@ -0,0 +1,586 @@
+
+def plot_cat_response_cat_predictor(
+    pred_data,
+    pred_name,
+    response,
+):
+    file_name = f"plots/{pred_name}-plot.html"
+    pred_data, _ = pd.factorize(pred_data)
+    response, _ = pd.factorize(response)
+    conf_matrix = confusion_matrix(pred_data, response)
+    fig = go.Figure(data=go.Heatmap(z=conf_matrix, zmin=0, zmax=conf_matrix.max()))
+    fig.update_layout(
+        title="Categorical Predictor by Categorical Response",
+        xaxis_title="Response",
+        yaxis_title=pred_name,
+    )
+    fig.write_html(
+        file=file_name,
+        include_plotlyjs="cdn",
+        include_mathjax="cdn",
+    )
+    return file_name
+    # fig.show()
+
+
+def plot_con_resp_cat_predictor(pred_data, pred_name, response):
+    file_name = f"plots/{pred_name}-plot.html"
+    fig = go.Figure()
+    for curr_hist, curr_group in zip(pred_data, response):
+        fig.add_trace(
+            go.Violin(
+                x=np.repeat(curr_group, 200),
+                y=[curr_hist],
+                name=curr_group,
+                box_visible=True,
+                meanline_visible=True,
+            )
+        )
+    fig.update_layout(
+        title="Continuous Response by Categorical Predictor",
+        xaxis_title=pred_name,
+        yaxis_title="Response",
+    )
+    fig.write_html(
+        file=file_name,
+        include_plotlyjs="cdn",
+        include_mathjax="cdn",
+    )
+
+    return file_name
+
+
+def plot_cat_resp_con_predictor(pred_data, pred_name, response):
+    file_name = f"plots/{pred_name}-plot.html"
+    group_labels = list(set([int(x) for x in response]))
+    hist_data0 = pred_data[response == group_labels[0]]
+    hist_data1 = pred_data[response == group_labels[1]]
+    hist_data = [hist_data0, hist_data1]
+    fig = ff.create_distplot(hist_data, group_labels, bin_size=0.2)
+    fig.update_layout(
+        title="Continuous Predictor by Categorical Response",
+        xaxis_title=pred_name,
+        yaxis_title="Distribution",
+    )
+
+    fig.write_html(
+        file=file_name,
+        include_plotlyjs="cdn",
+        include_mathjax="cdn",
+    )
+
+    return file_name
+
+
+def plot_con_response_con_predictor(pred_data, pred_name, response):
+    file_name = f"plots/{pred_name}-plot.html"
+    fig = px.scatter(x=pred_data, y=response, trendline="ols")
+    fig.update_layout(
+        title="Continuous Response by Continuous Predictor",
+        xaxis_title=pred_name,
+        yaxis_title="Response",
+    )
+
+    fig.write_html(
+        file=file_name,
+        include_plotlyjs="cdn",
+        include_mathjax="cdn",
+    )
+
+    return file_name
+
+
+def con_con_correlation(pred1_data, pred1_name, pred2_data, pred2_name):
+    correlation_stats, _ = scipy.stats.pearsonr(pred1_data, pred2_data)
+    return correlation_stats
+
+
+def con_cat_correlation(
+    pred1_data, pred1_name, pred1_type, pred2_data, pred2_name, pred2_type
+):
+    if len(set(pred1_data)) == 2 or len(set(pred2_data)) == 2:
```

Comment:
Nice using both biserial and correlation ratio whenever one is more appropriate (my system's currently only uses correlation ratio).